src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -58,28 +76,116 @@ public Task WriteDataAsync(ReadOnlySpan<byte> buffer, CancellationToken cancella
                 return Task.FromCanceled(cancellationToken);
             }
 
+            return WriteAsync(buffer, cancellationToken).AsTask();
+        }
+
+        public ValueTask<FlushResult> WriteDataToPipeAsync(ReadOnlySpan<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return new ValueTask<FlushResult>(Task.FromCanceled<FlushResult>(cancellationToken));
+            }
+
             return WriteAsync(buffer, cancellationToken);
         }
 
-        public Task WriteStreamSuffixAsync()
+        public ValueTask<FlushResult> WriteStreamSuffixAsync()
         {
             return WriteAsync(_endChunkedResponseBytes.Span);
         }
 
-        public Task FlushAsync(CancellationToken cancellationToken = default)
+        public ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
         {
             return WriteAsync(Constants.EmptyData, cancellationToken);
         }
 
-        public Task WriteChunkAsync(ReadOnlySpan<byte> buffer, CancellationToken cancellationToken)
+        public Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            if (_completed)
+            {
+                if (_completedMemoryOwner == null)
+                {
+                    _completedMemoryOwner = _memoryPool.Rent(sizeHint);
+                }
+                return _completedMemoryOwner.Memory;
+            }
+
+            if (_autoChunk)
+            {
+                return GetChunkedMemory(sizeHint);
+            }
+            else
+            {
+                return _pipeWriter.GetMemory(sizeHint);
+            }
+        }
+
+        public Span<byte> GetSpan(int sizeHint = 0)
+        {
+            if (_completed)
+            {
+                if (_completedMemoryOwner == null)
+                {
+                    _completedMemoryOwner = _memoryPool.Rent(sizeHint);
+                }
+                return _completedMemoryOwner.Memory.Span;
+            }
+
+            if (_autoChunk)
+            {
+                return GetChunkedMemory(sizeHint).Span;
+            }
+            else
+            {
+                return _pipeWriter.GetMemory(sizeHint).Span;
+            }
+        }
+
+        public void Advance(int bytes)
+        {
+            if (_completed)
+            {
+                return;
```

Comment:
Throw?

Reply:
Going to let this Noop still because PW.Complete can still be called by Abort. May revisit in a separate PR to distinguish calling Abort vs BodyPipe.Complete().

Reply:
Can you file an issue to track splitting Http1OutputProducer.Disable/Dispose()?

Reply:
Sure. I did some of it already here https://github.com/aspnet/AspNetCore/pull/7110/files#diff-cf42d74c2bbea44bc90926ce3dcee751R243

Reply:
What else do you think I need to do?